core-bundle/src/Security/Voter/GroupVoter.php

```diff
@@ -0,0 +1,42 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Security\Voter;
+
+use Contao\CoreBundle\Security\ContaoCorePermissions;
+use Contao\User;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\Voter;
+
+class GroupVoter extends Voter
+{
+    protected function supports($attribute, $subject): bool
+    {
+        if (ContaoCorePermissions::MEMBER_OF_GROUPS !== $attribute) {
+            return false;
+        }
+
+        return is_numeric($subject)
+            || (\is_array($subject) && \count($subject) === \count(array_filter($subject, 'is_numeric')));
+    }
+
+    protected function voteOnAttribute($attribute, $subject, TokenInterface $token): bool
+    {
+        $user = $token->getUser();
+
+        if (!$user instanceof User) {
+            return false;
```

Comment:
Couldn’t we check the `-1` group here so that we don’t have to do that in all the other places?

```suggestion
            return in_array(-1, array_map('intval', (array) $subject), true);
```

Reply:
Thats not the same at all. Calling `is this user in group 1,2,3` shouldn't return true if I don't have a user (and not checking the groups at all). Being a guest means not having `ROLE_MEMBER`, not _not_ having a Contao user.

Reply:
Should it be something like this then?

```php
if (!$user instanceof User || !\in_array('ROLE_MEMBER', $token->getRoleNames(), true)) {
    return in_array(-1, array_map('intval', (array) $subject), true);
}
```

Reply:
Handling a magic group `-1` is wrong imho in the first place, that's what I don't like about the change. As I said in #3088 I feel like that's gonna do more harm than good. But thats not related to this PR at all.

Anyway, calling `isGranted('contao_group', -1)` and expecting the system to return `true` if there's no user is wrong imho. Also, this voter is not supposed to vote on role names, it only votes on the known attribute. Voting on roles is done by the role voter, which you might override for any other reason.

Reply:
OK, I see. So this voter is used for both, users (backend) and members (frontend)?

Wouldn’t it make sense to separate it into two voters, so that the voter for members can handle the `-1` case and the voter for users not?

Reply:
The voters we currently have vote on features we currently have. The `BackendAccessVoters` votes on `BackendUser::isAllowed` etc, the `GroupVoter` votes on `User::isMemberOf`.

I think what you're asking for would be a completely separate voter, and I'm not sure that's intended by the Symfony security system. Voters vote on one specific attribute, a voter is not supposed to make a final decision for multiple conditions (e.g. a user is not logged in AND an element is set to be visible to a guest only). It sure would work if we agree that would be necessary, but then it wouldn't be a different voter concept.

Reply:
I still think we should split it up in a `BackendGroupVoter` and a `FrontendGroupVoter`, otherwise if someone forgets to vote on the role beforehand, the `GroupVoter` would return true if the IDs of the backend groups accidentaly match with the IDs of the frontend groups.

Also, to know if a frontend group matches, we have to do several checks instead of just one:

1. Check if there is a user
2. If not check for the `-1` group
3. Ask the voter for `ROLE_MEMBER`
4. If not check for the `-1` group
5. Ask the voter for the group

Reply:
I second this.

Reply:
again, it will no longer be a group voter then. It will be a sort of "has access voter". Also, again, checking if there is user is not done by checking an interface but for the role, and thats not supposed to be in this voter.

Reply:
> It will be a sort of "has access voter".

Similar to the `BackendAccessVoters` should we name them `BackendGroupAccessVoter` and `FrontendGroupAccessVoter` then?